app/services/get-scenario.js

```diff
@@ -0,0 +1,323 @@
+const config = require('config');
+const qs = require('qs');
+const db = require('../db');
+const _ = require('lodash');
+
+// Create cache client only if Redis is available
+// Get Redis config
+const cache = config.get('cache');
+let redisClient;
+if (cache.enabled) {
+  redisClient = require('../redis-client');
+}
+
+/**
+ * Generate a string key for a scenarioId/query pair.
+ *
+ * @param {string} id Scenario id.
+ * @param {string} query Query
+ */
+function cacheKeyFromQuery (id, query) {
+  return JSON.stringify({ id, query });
+}
+
+async function prepareQuery (id, querystring) {
+  let filters;
+  let targetYear;
+  const modelId = id.substring(0, id.lastIndexOf('-'));
+
+  if (querystring) {
+    let query = qs.parse(querystring);
+    targetYear = parseInt(query.year) || null;
+
+    // Validate and parse filters
+    filters = query.filters;
+    if (filters) {
+      // Filters must be in an Array
+      if (!Array.isArray(filters)) {
+        throw new SyntaxError('Filters must be an Array.');
+      }
+
+      // Validate range values
+      filters.forEach(filter => {
+        // A filter key must be defined
+        if (typeof filter.key === 'undefined') {
+          throw new SyntaxError('Filter must include "key".');
+        }
+
+        if (
+          typeof filter.min === 'undefined' &&
+          typeof filter.max === 'undefined' &&
+          typeof filter.options === 'undefined'
+        ) {
+          throw new SyntaxError(
+            'Filter must include a valid value parameter name: "min", "max" or "options").'
+          );
+        }
+      });
+    }
+  }
+
+  // Get information about the model
+  const model = await db('models')
+    .select('filters', 'timesteps', 'baseYear')
+    .where('id', modelId)
+    .first();
+
+  // Validate timestep model
+  if (model.timesteps) {
+    // Set the year to the 1st value if null
+    targetYear = targetYear || model.timesteps[0];
+    if (model.timesteps.indexOf(targetYear) === -1) {
+      throw new SyntaxError(
+        `The "year" parameter [${targetYear}] is invalid for this scenario. Must be one of [${model.timesteps.join(
+          ', '
+        )}]`
+      );
+    }
+  } else {
+    // Disregard year
+    targetYear = '';
+  }
+
+  return {
+    id,
+    model,
+    filters,
+    querystring,
+    targetYear
+  };
+}
+
+async function lockQuery (id, querystring) {
+  // Generate cache key from query params
+  const cacheKey = cacheKeyFromQuery(id, querystring);
+
+  // Set cache flag "runningDbQuery" as true
+  if (cache.enabled) {
+    await redisClient.setObject(
+      cacheKey,
+      { runningDbQuery: true },
+      cache.loadTimeout
```

Comment:
The lock timeout must be different than the load timeout.
It may happen that the load times out and the query is still running.

The lock timeout should be a high number (like 15m) which will be cleared upon success or failure of the query operation.